packages/extension/src/shared/transactions/index.ts

```diff
@@ -129,3 +133,295 @@ function buildGetTransactionsFn(metadataTransactions: Transaction[]) {
+            return signedChainedTxResultToReviewTransactionResult(params, chainedTxResults[index], account.networkId)
+          })
+
+          return reviewTxResults
+        }
+      }
+
+      const [firstMissingTokenId, firstMissingAmount] = missingBalances.entries().next().value;
+      const tokenSymbol = firstMissingTokenId === ALPH_TOKEN_ID ? 'ALPH' : firstMissingTokenId;
+      const expectedStr = firstMissingAmount.toString();
+      const haveStr = (tokensWithBalance.find(t => t.id === firstMissingTokenId)?.balance || '0').toString();
+      const errorMsg = i18n.t("Insufficient token {{ tokenSymbol }}, expected at least {{ expectedStr }}, got {{ haveStr }}", { tokenSymbol, expectedStr, haveStr })
+      throw new Error(errorMsg)
+    }
+
+    throw error
+  }
+}
+
+async function accountWithEnoughBalance(
+  missingBalances: Map<string, BigNumber>,
+  accounts: WalletAccount[],
+  nodeProvider: NodeProvider
+): Promise<WalletAccount | undefined> {
+  for (const account of accounts) {
+    const accountBalances = await getBalances(nodeProvider, account.address)
+    let hasEnoughBalance = true;
+    for (const [tokenId, amount] of missingBalances) {
+      const balance = accountBalances.get(tokenId);
+      if (!balance || balance.lt(amount)) {
+        hasEnoughBalance = false;
+        break;
+      }
+    }
+
+    if (hasEnoughBalance) {
+      return account
+    }
+  }
+  return undefined
+}
+
+
+async function buildTransaction(
+  builder: TransactionBuilder,
+  account: WalletAccount,
+  transactionParams: TransactionParams
+): Promise<ReviewTransactionResult> {
+  switch (transactionParams.type) {
+    case "TRANSFER":
+      return {
+        type: transactionParams.type,
+        params: transactionParams.params,
+        result: await builder.buildTransferTx(
+          transactionParams.params,
+          account.signer.publicKey,
+        ),
+      }
+    case "DEPLOY_CONTRACT":
+      return {
+        type: transactionParams.type,
+        params: transactionParams.params,
+        result: await builder.buildDeployContractTx(
+          transactionParams.params,
+          account.signer.publicKey,
+        ),
+      }
+    case "EXECUTE_SCRIPT":
+      return {
+        type: transactionParams.type,
+        params: transactionParams.params,
+        result: await builder.buildExecuteScriptTx(
+          transactionParams.params,
+          account.signer.publicKey,
+        ),
+      }
+    case "UNSIGNED_TX":
+      return {
+        type: transactionParams.type,
+        params: transactionParams.params,
+        result: await builder.buildUnsignedTx(transactionParams.params),
+      }
+  }
+}
+
+export async function checkBalances(
+  tokensWithBalance: BaseTokenWithBalance[],
+  transactionParams: TransactionParams
+): Promise<Map<string, BigNumber> | undefined> {
+  const expectedBalances: Map<string, BigNumber> = new Map()
+  const gasFee = BigInt(5000000) * DEFAULT_GAS_PRICE  // Maximal gas fee per tx
```

Comment:
If the user specifies both `gasAmount` and `gasPrice`, maybe we could use `gasAmount * gasPrice`. WDYT?